Support :semi_join_first option when inner joining

jeremyevans opened the pull request:
This runs a LEFT SEMI JOIN between the two tables, which will remove
rows in the left hand side that are not matching in the right hand
side, use that as a subquery and then run an INNER JOIN on that subquery
and the original right hand side.

This will not work correctly if there are multiple tables being joined
and the join call using the :semi_join_first option uses an expression
with a qualified reference to a table that isn't the first table or
the table being joined to.  For example, the following will not work:

  DB[:a].join(:b, :a_id=>:id).
    join(:c, {:id=>Sequel[:b][:id]}, semi_join_first: true)

this is because the reference to b.id is not valid as b is in the
LEFT SEMI JOIN subquery in the left hand side, and that cannot be
referenced outside the subquery during the INNER JOIN.

That could theoretically be worked around by replacing all qualified
references in the join expression that do not match the first table
alias in the query or the table being joined with the the alias of
the subquery.

jeremyevans committed:
Support :semi_join_first option when inner joining

This runs a LEFT SEMI JOIN between the two tables, which will remove
rows in the left hand side that are not matching in the right hand
side, use that as a subquery and then run an INNER JOIN on that subquery
and the original right hand side.

This will not work correctly if there are multiple tables being joined
and the join call using the :semi_join_first option uses an expression
with a qualified reference to a table that isn't the first table or
the table being joined to.  For example, the following will not work:

  DB[:a].join(:b, :a_id=>:id).
    join(:c, {:id=>Sequel[:b][:id]}, semi_join_first: true)

this is because the reference to b.id is not valid as b is in the
LEFT SEMI JOIN subquery in the left hand side, and that cannot be
referenced outside the subquery during the INNER JOIN.

That could theoretically be worked around by replacing all qualified
references in the join expression that do not match the first table
alias in the query or the table being joined with the the alias of
the subquery.

aguynamedryan commented:
Thanks, @jeremyevans! 

Is it possible to also make it so that the results of the LEFT SEMI JOIN are stored in a "temporary" table?  I'm concerned that simply making it a subquery will still somehow befuddle Impala and exhaust its resources before the query completes.  Breaking it into two discrete steps using a "temporary" table has been helpful in avoiding that in the past.  I will need to be able to pass in the name of the database in which this temporary table should be stored, and the temporary table name needs to be unique for each call so that concurrent processes don't end up using the same name for the temporary table.

If it's not _terribly_ hard, I'd like the option of either keeping the code you have now where the LEFT SEMI JOIN is a subquery, and another option of writing the results of the LEFT SEMI JOIN to a temporary table first.

jeremyevans commented:
@aguynamedryan Thanks for the quick feedback, I can work on that right away.

jeremyevans committed:
Support the :semi_join_first option being a temporary table name

If the :semi_join_first option is given as a table name (either
String, Symbol, SQL::Identifier, or SQL::QualifiedIdentifier),
then instead of using a subquery, select from that table name.
Then when executing the query, create any necessary :semi_join_first
tables before the query execution, and drop the tables after the
query execution.

jeremyevans commented:
OK, this last commit allows you to use `semi_join_first: true` for the subquery approach, and `semi_join_first: :table_name` or `semi_join_first: Sequel[:impala_database_name][:table_name]` to use a temporary table for the execution.

This pushes the generation of the temporary table name to the calling code, but I expect the calling code in this case already has easy access to a temporary table name generator.

aguynamedryan commented:
@jeremyevans, thank you very much!
